write-fonts/src/pens.rs

```diff
@@ -0,0 +1,369 @@
+pub struct ReverseContourPen<'a, T: Pen> {
+    inner_pen: &'a mut T,
+    pending: Vec<PenCommand>,
+}
+
+fn end(subpath_start_x: f32, subpath_start_y: f32, subpath_cmd: PenCommand) -> (f32, f32) {
+    match subpath_cmd {
+        PenCommand::MoveTo { x, y } => (x, y),
+        PenCommand::LineTo { x, y } => (x, y),
+        PenCommand::QuadTo { x, y, .. } => (x, y),
+        PenCommand::CurveTo { x, y, .. } => (x, y),
+        PenCommand::Close => (subpath_start_x, subpath_start_y),
+    }
+}
+
+/// Reverse the commands in a path.
+///
+///  FontTools version in
+/// <https://github.com/fonttools/fonttools/blob/78e10d8b42095b709cd4125e592d914d3ed1558e/Lib/fontTools/pens/reverseContourPen.py#L25>
+fn flush_subpath<T: Pen>(commands: &[PenCommand], pen: &mut T) -> Result<(), ContourReversalError> {
+    if commands.is_empty() {
+        return Ok(());
+    }
+
+    let mut commands = commands;
+    let mut reversed = Vec::new();
+
+    // subpath by definition contains 0 or 1 move commands, and the move is at [0] if it exists
+    let (start_x, start_y) = if let PenCommand::MoveTo { x, y } = commands[0] {
+        // When reversed, the move is to the end point of the last command
+        // in a typical [move, ..., close] structure, end point == start point
+        let (end_x, end_y) = end(x, y, *commands.last().unwrap());
+        reversed.push(PenCommand::MoveTo { x: end_x, y: end_y });
+        commands = &commands[1..];
+        (x, y)
+    } else {
+        (0.0, 0.0)
+    };
+
+    // Reverse the commands between move (if any) and final close (if any)
+    for (idx, cmd) in commands.iter().enumerate().rev() {
+        let (end_x, end_y) = if idx > 0 {
+            end(start_x, start_y, commands[idx - 1])
+        // } else if ends_with_close {
+        //     (start_x, start_y)
+        } else {
+            (start_x, start_y)
+        };
+        reversed.push(match *cmd {
+            PenCommand::MoveTo { .. } => {
+                panic!("Subpath should have 0 or 1 moves, and it should already have been removed")
+            }
+            PenCommand::LineTo { .. } => PenCommand::LineTo { x: end_x, y: end_y },
+            PenCommand::QuadTo { cx0, cy0, .. } => PenCommand::QuadTo {
+                cx0,
+                cy0,
+                x: end_x,
+                y: end_y,
+            },
+            PenCommand::CurveTo {
+                cx0, cy0, cx1, cy1, ..
+            } => PenCommand::CurveTo {
+                cx0: cx1,
+                cy0: cy1,
+                cx1: cx0,
+                cy1: cy0,
+                x: end_x,
+                y: end_y,
+            },
+            // Close is a line from (end_x,end_y)=>(start_x, start_y) so reversed it's a line to end x/y
+            PenCommand::Close => PenCommand::LineTo { x: end_x, y: end_y },
+        });
+    }
+
+    // a closing line to start is a Z
```

Comment:
what you're doing here (replacing the last LineTo of the reversed path with a ClosePath when it overlaps the starting point) is problematic because you may inadvertently drop a duplicate lineTo point that immediately followed moveTo pre-reversal and which deliberately has the same (x,y) coordinates, thus potentially making VF sources incompatible after reversing.

In fontTools ReverseContourPen (as well as in the PointToSegmentPen that converts between points and segment-oriented pens) we have an option outputImpliedClosingLine that when True it always output the implied closing line regardless of whether the last point overlaps moveTo or not. Setting this option is preferable when working with VF masters.
eg. see this font where some contour points collaps onto one another but only in some masters https://github.com/googlefonts/fontmake/issues/572

But even when outputImpliedClosingLine=False, fonttools ReverseContourPen will _not_ omit the last lineTo (which was the first one, before reversing) if it overlaps with the starting point, otherwise the number of points and segments would change after reversing
See https://github.com/googlefonts/cu2qu/issues/51 (fixed by https://github.com/fonttools/fonttools/pull/1080)

see this test case: 
https://github.com/fonttools/fonttools/blob/bf265ce49e0cae6f032420a4c80c31d8e16285b8/Tests/pens/reverseContourPen_test.py#L227-L250

and also this one: https://github.com/fonttools/fonttools/blob/bf265ce49e0cae6f032420a4c80c31d8e16285b8/Tests/pens/reverseContourPen_test.py#L38-L53

actually _all_ the tests in reverseContourPen_test.py would be nice to port over...

Reply:
Good call re porting tests, will do.

Reply:
I'm dubious about the FontTools practice of eliminating the line from a close during reversal. Imagine you have several versions of move/line/line/close and in _some_ of them the last line ends at the start ... if you eliminate the "unnecessary" line from close as suggested by https://github.com/fonttools/fonttools/blob/bf265ce49e0cae6f032420a4c80c31d8e16285b8/Tests/pens/reverseContourPen_test.py#L23 - note that an input MLLLC turns into an output MLLC - then these shapes will no longer interpolate.

Reply:
I suppose ^ is to say I like the outputImpliedClosingLine=True behavior and do not currently implement the FontTools default mode of dropping the closing lines.